Stratis.Bitcoin/BlockStore/BlockStoreBehavior.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NBitcoin;
+using NBitcoin.Protocol;
+using NBitcoin.Protocol.Behaviors;
+using Stratis.Bitcoin.Consensus;
+
+namespace Stratis.Bitcoin.BlockStore
+{
+	public class BlockStoreBehavior : NodeBehavior
+	{
+		private readonly ConcurrentChain concurrentChain;
+		private readonly BlockRepository blockRepository;
+
+		public bool CanRespondToGetBlocksPayload { get; set; }
+
+		public bool CanRespondeToGetDataPayload { get; set; }
+
+		public BlockStoreBehavior(ConcurrentChain concurrentChain, BlockRepository blockRepository)
+		{
+			this.concurrentChain = concurrentChain;
+			this.blockRepository = blockRepository;
+
+			this.CanRespondToGetBlocksPayload = false;
+			this.CanRespondeToGetDataPayload = true;
+		}
+
+		protected override void AttachCore()
+		{	
+			this.AttachedNode.MessageReceived += AttachedNode_MessageReceived;
+		}
+
+		protected override void DetachCore()
+		{
+			this.AttachedNode.MessageReceived -= AttachedNode_MessageReceived;
+		}
+
+		private async void AttachedNode_MessageReceived(Node node, IncomingMessage message)
+		{
+			await this.AttachedNode_MessageReceivedAsync(node, message);
+		}
+
+		private async Task AttachedNode_MessageReceivedAsync(Node node, IncomingMessage message)
+		{
+			if (this.CanRespondeToGetDataPayload)
+				await message.Message.IfPayloadIsAsync<GetDataPayload>(async data => await this.ProcessGetDataAsync(node, data));
```

Comment:
when you do a await and do not expect the result to be queued on the default scheduler, you need to do: `await x.ConfigureAwaiter(false)`